core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -989,17 +725,501 @@ private static SortOrder freshSortOrder(int orderId, Schema schema, SortOrder so
+
+      return this;
+    }
+
+    public Builder setCurrentSnapshot(Snapshot snapshot) {
+      addSnapshot(snapshot);
+      setCurrentSnapshot(snapshot, snapshot.timestampMillis());
+      return this;
+    }
+
+    private Builder setCurrentSnapshot(long snapshotId) {
+      if (currentSnapshotId == snapshotId) {
+        // change is a noop
+        return this;
+      }
+
+      Snapshot snapshot = snapshotsById.get(snapshotId);
+      ValidationException.check(snapshot != null,
+          "Cannot set current snapshot to unknown: %s", snapshotId);
+
+      setCurrentSnapshot(snapshot, System.currentTimeMillis());
+
+      return this;
+    }
+
+    public Builder removeSnapshots(List<Snapshot> snapshotsToRemove) {
+      Set<Long> idsToRemove = snapshotsToRemove.stream().map(Snapshot::snapshotId).collect(Collectors.toSet());
+
+      List<Snapshot> retainedSnapshots = Lists.newArrayListWithExpectedSize(snapshots.size() - idsToRemove.size());
+      for (Snapshot snapshot : snapshots) {
+        long snapshotId = snapshot.snapshotId();
+        if (idsToRemove.contains(snapshotId)) {
+          snapshotsById.remove(snapshotId);
+          changes.add(new MetadataUpdate.RemoveSnapshot(snapshotId));
+        } else {
+          retainedSnapshots.add(snapshot);
+        }
+      }
+
+      this.snapshots = retainedSnapshots;
+      if (!snapshotsById.containsKey(currentSnapshotId)) {
+        setCurrentSnapshot(null, System.currentTimeMillis());
+      }
+
+      return this;
+    }
+
+    public Builder setProperties(Map<String, String> updated) {
+      if (updated.isEmpty()) {
+        return this;
+      }
+
+      properties.putAll(updated);
+      changes.add(new MetadataUpdate.SetProperties(updated));
+
+      return this;
+    }
+
+    public Builder removeProperties(Set<String> removed) {
+      if (removed.isEmpty()) {
+        return this;
+      }
+
+      removed.forEach(properties::remove);
+      changes.add(new MetadataUpdate.RemoveProperties(removed));
+
+      return this;
+    }
+
+    public Builder setLocation(String newLocation) {
+      if (location != null && !location.equals(newLocation)) {
+        return this;
       }
+
+      this.location = newLocation;
+      changes.add(new MetadataUpdate.SetLocation(newLocation));
+
+      return this;
+    }
+
+    public TableMetadata build() {
+      if (changes.size() == startingChangeCount) {
+        return base;
+      }
+
+      if (lastUpdatedMillis == null) {
+        this.lastUpdatedMillis = System.currentTimeMillis();
+      }
+
+      Schema schema = schemasById.get(currentSchemaId);
+      PartitionSpec.checkCompatibility(specsById.get(defaultSpecId), schema);
+      SortOrder.checkCompatibility(sortOrdersById.get(defaultSortOrderId), schema);
+
+      List<MetadataLogEntry> metadataHistory = addPreviousFile(previousFileLocation, base.lastUpdatedMillis());
+
+      // find intermediate snapshots to suppress incorrect entries in the snapshot log.
```

Comment:
This is new. It identifies intermediate snapshots that were added and set as current without being committed. The logs for these are suppressed.